### DISCUSSION TITLE
Set relationMode per table or per schema

### DISCUSSION THREAD

**Discussion by olivierlacan** [prisma/prisma#21980] [Q&A] [answered]
### Question

Our team uses database tables that we'd like to query with Prisma but they're streamed into our Postgres database (in a separate schema than `public`) by a process we don't own. This system can't rely on foreign key constraints but it would extremely helpful for us to be able to assign the `relationMode` to that specific schema in our `datasource`.

We really don't want to enable this for our `public` schema, where foreign key constraints are ideal. 

In practice this could look like this: 

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["ext", "public"]
  relationMode = [{ "ext": "prisma" }, { "public": "foreignKey" }]
}
```

Now I'm hand-waving the implications for the Prisma client of course, but that would allow us exactly the right amount of referential integrity we need (or can afford) without sacrificing it for the tables we own which do leverage foreign key constraints. 

Does that make any sense? 

PS: I'm leaving aside the more complex issue of having to create relations from one relationMode=foreignKey schema to another schema set to relationMode=schema for another day. That seems like pushing it. 😅

**Comment by janpio** [comment, marked as answer]
As you know this is currently not possible. I am pretty sure we have an issue for it somewhere, but could not quickly find it. I can have another look later.

But you can probably work around that.

`relationMode=prisma` really does 2 things: 1. It makes sure no foreign keys are created in the SQL of migrations, and 2. that the missing functionality the foreign keys would bring, like referential actions and referential integrity, are handled by Prisma itself. You should be able to use the normal Prisma relationMode, remove the SQL migrations for `ext` from the migration files before executing them (and let the external system define the schemas) or modify them to not have foreign keys. Then, as you are in the default relationMode, you only need to yourself manage the "foreign key columns" by potentially doing some additional queries for cleanup. That way, you _should_ be able to use Prisma with the current setup. It will take some thinking and potentially additional code or queries, but Prisma will be ok with that situation.

**Reply by olivierlacan** [reply]
Thanks for the prompt response! 

We ended up enabling `relationMode=prisma` for now then using the helpful @relation hints to ensure indices were defined for all those relations. Is there a way to configure those hints to be errors perhaps? That would be a nice guardrail against index-free relations, at least as opt-in feature. 

Being in `foreignKey` mode was a bit tricky since `prisma migrate dev` would read the @relation established on the `ext` models and attempt to create foreign key constraints on them since some are inter-related. That's fairly dangerous since a developer could accidentally establish foreign key constraints on two tables in `ext`.

Similarly any @relation from a `public` table to an `ext` one could lead to a foreign key constraint being created in `public` that would prevent deletion of `ext` rows, which is too risky.

**Reply by janpio** [reply]
Yes, going with `relationMode=foreignKey` would have indeed meant watching the generated migrations to make sure not bad SQL is introduced to the system. I understand that this is too much a risk, and too much work to prevent via tooling.

Right now there is no way to turn the index warnings into errors. You might want to create a feature request (but I'll already let you know that we are not going to jump on this, as we are currently trying to avoid to create configuration as much as possible. This will change in the future.)
